### DISCUSSION TITLE
copilot.vim support for copilot chat?

### DISCUSSION THREAD

**Discussion by zdufour-asp** [community/community#50939] [Copilot Conversations]
### Select Topic Area

Product Feedback

### Body

Will people in the technical preview for copilot x have direct access to an API?

I'm assuming that there won't be `copilot.vim` support during the preview, but I'd be down to implement a hacky UI for neovim if it's possible (I really want to try out the new tech, but having to set up the vscode neovim plugin to do so seems like a lot of work)

**Comment by Akash1134** [comment]
Hi @zdufour-asp , Thank you for your feedback and the excitement for the new tech 🚀 . Since the product is in the public preview, it is pretty early for us to comment on future developments.  [Sign up for the private preview](https://aka.ms/GHCopilotXPPSignup) now if you haven't so far. 🙂

**Reply by DanielPower** [reply]
Now that Copilot Chat is available for all organizations[^1], is there any update on whether it will be implemented in copilot.vim?

[^1]: https://github.blog/2023-07-20-github-copilot-chat-beta-now-available-for-every-organization/

**Comment by giner** [comment]
> [Sign up for the private preview](https://aka.ms/GHCopilotXPPSignup) now if you haven't so far

There are no options vim or "other" on signing up

**Reply by MG72021** [reply]
[google.com](url)

**Comment by chgara** [comment]
Now that Github Copilot Chat is open for all, will be developed a pulgin for neovim?

**Comment by IzioDev** [comment]
Currently paying 100$ per year for basically `Copilot`, now it's late 2023, I wish I could benefit the most from this expense, but I cannot as I'm a Neovim user.

re-installing VsCode would break my heart and my colleagues would laugh at me, out of the equation.

**Reply by scnew** [reply]
Hi IzioDev: I am a vim user and have not been able to get nodejs + npm working on Cygwin + Windows.   I guess I am confused.  To me it is not clear if vim running on Cygwin (Windows) is supported by copliot.  I have left a few posts and not heard back.  Have you seen any install info?  Kind Regards, and thanks for response in advance.  e

**Reply by jack-michaud** [reply]
@IzioDev I am in the same boat. I have a vim tattoo, I can't back out now.

**Comment by georgeanderson** [comment]
Copilot Chat seems to be rolling out for more people. Any plans for adding support to neovim?

**Comment by howarddo2208** [comment]
looking forward to this!

**Comment by aramirezreyes** [comment]
One more vote for getting Copilot chat support on the vim

**Comment by faraazkhan** [comment]
Another vote for maintaining general parity between the vim and vscode versions

**Comment by bmulholland** [comment]
Folks: There's a vote feature at the top. Using that adds your support, without notifying anyone. I'd like to stay subscribed to this, so I can get actual updates, but when you add a comment that's just "another vote," you achieve nothing except annoying me and everyone else who is subscribed. If people keep doing it, I'll have to unsubscribe, and then I won't stay in the loop. So please, don't comment with a +1.

**Reply by JohnRPB** [reply]
Reading the Github docs, it's not completely clear to me when an issue will close as stale, and whether upvotes alone count as activity. If they do, then we should all follow your advice.

**Reply by wootfish** [reply]
I'm in favor of adding more upvotes, but I will note that this topic already is the #1 all-time most voted-for issue, with almost three times as many votes as the next unanswered topic. If votes do matter, one has to wonder how many more votes would be needed to get a response.

**Reply by gshpychka** [reply]
>  Reading the Github docs, it's not completely clear to me when an issue will close as stale, and whether upvotes alone count as activity. If they do, then we should all follow your advice.

This is not an issue, it's a discussion, so it won't be closed as stale.

> I'm in favor of adding more upvotes, but I will note that this topic already is the https://github.com/orgs/community/discussions/1 all-time most voted-for issue, with almost three times as many votes as the next unanswered topic. If votes do matter, one has to wonder how many more votes would be needed to get a response.

This is just up to Microsoft at this point, spamming this discussion with comments is very unlikely to affect their decision making any more than the upvotes.

**Comment by weykon** [comment]


**Reply by bmulholland** [reply]
https://github.com/orgs/community/discussions/50939#discussioncomment-7402566

**Reply by aramirezreyes** [reply]
Sorry @bmulholland, now I am getting two notifications, one with the comment and one with you mentioning how annoyed you are by the comment. Are you a moderator or trying to enforce community guidelines or just trying to enforce your personal preference? (I could not find where it is specified that the way to interact is the button instead of a comment, could you kindly point me to it?). 

Are you part of github and would be involved in the implementation (or as a moderator), or just another user like the rest of us commenting?

**Reply by gshpychka** [reply]
@aramirezreyes it's just common github etiquette. Community support is gauged by the votes on the topic, not by (arguably redundant) comments (each one sending notifications to everyone who is subscribed to the discussion)

**Reply by IzioDev** [reply]
at the end of the day, adults are just like child but arguing with more words. funny, isn't it? :D

**Comment by anon-legion** [comment]
would love this feature to be ported to nvim as well

**Comment by JohnRPB** [comment]
When Copilot says "you can work in your favorite editor", and then preferentially rolls out features to only the Microsoft-connected editors for half a year, offering no progress updates on support for other editors, it looks super sus.

**Reply by FlippingBinary** [reply]
Any publicly traded company should be expected to prioritize adding features to their own tools over others. That doesn't mean they won't follow through. Imagine the blowback if they had Copilot Chat working in NeoVim while VS Code users have to wait six months or more to start using it. The development timelines are different because a NeoVim plugin is a totally separate product from a VS Code plugin, so a synchronized release would inevitably result in a delay for one or the other.

Luckily there are alternatives to copilot, if you're interested. [model.nvim](https://github.com/gsuuon/model.nvim) lets you connect with a variety of AI models for code completion and chat. You can even connect it to a local LLM, which is nice if you ever work on anything proprietary.

**Reply by peedrr** [reply]
> preferentially rolls out features to only the Microsoft-connected editors

Imagine they were your products, your IDE (the key word here being "Integrated"), your Dev budget... we're all keen for the full Copilot experience in Neovim but to develop, prove and perfect a new product outside of your own, existing ecosystem would quite frankly be a ridiculous business decision. I'm sure they're keeping an eye on external integrations - again, a good business strategy because of the licensing model - but it wouldn't make sense to prioritise them

**Reply by JohnRPB** [reply]
I think these are good points, and I have no problem with some prioritization of Microsoft-connected tooling, but I do have a problem with the "no progress updates for six months" part. I started using Copilot because it said it had first-class support for editors including Neovim, which seemed to signal that Github was making a good faith effort to maintain open-source compatibility, which was a big concern when Github was acquired.

Now I'm sitting here, after a year of Copilot X hype, feeling sorely tempted to use VSCode with the neovim plugin, and it all feels a bit more like a ploy to increase market share. Or, as with a lot of these things, there's genuinely good intentions behind it, but prioritization slips so much under the publicly-traded profit motive that it becomes indistinguishable from malice.

Is it right around the corner? Six months away? Just asking for a genuine update, so I can make an informed decision about switching to model.nvim or something similar.

**Reply by topocount** [reply]
It's also worth noting that this is far and away the most requested feature for copilot, which makes the lack of updates all the more puzzling.

**Comment by jdstrand** [comment]
Fyi, a community person wrote https://github.com/gptlang/CopilotChat.nvim. I read it (checkout of [798a48b](https://github.com/gptlang/CopilotChat.nvim/commit/798a48b0a84964da91b526bbc6ab48f111e7396f)) and it is rough around the edges, but the basic idea is that it provides a `:CopilotChat` command that simply takes a prompt (eg, `:CopilotChat hi there!`). This will open a scratch file then set `:setlocal buftype=nofile bufhidden=hide noswapfile`. I played with it a bit today and found this to be fairly usable:

1. open a file in nvim
2. do `:vsplit` followed by `:enew` followed by `:setlocal buftype=nofile bufhidden=hide noswapfile wrap linebreak nonu` to open a scratch file with pleasant wrapping, etc
3. start chatting `:CopilotChat hi there!`
4. go to another buffer (eg, the file you opened with) and yank (`y`) a function into a buffer. Go back to the scratch file (a limitation of the `CopilotChat.nvim` is it will open a new scratch file if the current buffer is not one)
5. tell it to do stuff: `:CopilotChat explain this function` or `:CopilotChat write documentation for this function` or `:CopilotChat write unit tests for this function` or `:CopilotChat how can this function be made better:` or ...
6. copy (yank) and paste from the scratch buffer into other buffers to incorporate the changes

I mention this not so much to highlight this project (I don't have anything to do with it), but to hopefully inspire the `copilot.vim` devs and share a bit of how a workflow could work with new functionality in `copilot.vim`. I imagine `copilot.vim`. could have configurable settings for a vertical vs horizontal split, using a named buffer rather than a scratch file, being able to input a chat prompt from any buffer and have it show up in the chat (named buffer), etc, etc.

PS - if people do play with it, I found I needed python 3.10 (3.8 didn't work) and needed to set PYTHONPATH to `$HOME/.config/nvim/rplugin/python3` to use.

**Comment by georgeanderson** [comment]
I have no horse in this race, but maybe Codeium could release something like chat for Neovim first and that would be interesting on their adoption side over Copilot. I use both in Neovim and they produce very similar outcomes.

**Comment by kkzhang** [comment]
vim support for chat +1

**Reply by gshpychka** [reply]
https://github.com/orgs/community/discussions/50939#discussioncomment-7402566

**Comment by daufraud** [comment]
+1

**Reply by gshpychka** [reply]
https://github.com/orgs/community/discussions/50939#discussioncomment-7402566

**Comment by wootfish** [comment]
As a committed Vim user I have to admit Copilot's advertising around Vim support feels deceptive. It is implied in marketing materials that Vim users will have access to the same features as users of other platforms, when in fact one of the biggest selling points of Copilot is completely missing. There is no timeline for chat, or even an indication that chat will be added.

Users could not be blamed for assuming the Vim plugin was a marketing gimmick which has served its purpose and been abandoned. No one wants this, and if this is not the case I hope we can get some indication to the contrary soon.

**Reply by johnknl** [reply]
As someone old enough to remember developing for IE6, I have an uncomfortable feeling about MS shenanigans like this. Seems like the main lesson they learned is to not only bind users but also developers to them and give them as little choice in the matter as possible.

**Reply by DanielPower** [reply]
Vim support is one of the reasons I pushed for Copilot over other options when my company was trialing AI tools. I suspect I wasn't the only vocal Vim user doing this at their respective companies. 

Now that Microsoft has achieved a significant market share, it feels like they no longer need to care about Vim and that we were misled for marketing / market penetration purposes.

**Reply by atsiporu** [reply]
Same here. I regret the push now.

**Comment by llucax** [comment]
Now it is "generally" available... :angry: 

* https://github.com/orgs/community/discussions/83931

> All GitHub Copilot users can now enjoy natural language-powered coding with Copilot Chat at no additional cost.

"All". "No extra cost", it is quite costly to switch IDEs. It is adevertised to work Neovim, this is just outrageous and unacceptable.

**Comment by llucax** [comment]
Maybe we should just get a beach house...

[![image](https://github.com/community/community/assets/1031485/11b7f169-b83b-4e96-bad8-f3973859fc65)](https://xkcd.com/2881/)

**Comment by mishmanners** [comment]
You can always check the GitHub Public Road map to see what's coming. Here's the query for [searching upcoming GitHub Copilot features](https://github.com/orgs/github/projects/4247/views/1?filterQuery=is%3Aopen+copilot).

**Reply by wootfish** [reply]
This is helpful, thank you. When I view this page I see one item (an API deprecation) listed for Jan-Mar 2024 and no items for any other time period.

Judging by what I am seeing, it would appear that no further work on the Vim plugin is planned, but I am a bit confused as this does not seem to align with your comment in the Copilot Chat announcement thread:

> Yes GitHub Copilot is available in NeoVim, but not GitHub Copilot Chat... that should be coming soon. Keep an eye out.

Is Github Copilot Chat actually coming soon for Vim/NeoVim? If so, how soon?

**Reply by gshpychka** [reply]
Even the JetBrains integration going GA isn't on the roadmap.

**Reply by llucax** [reply]
@mishmanners I don't want to be rude, but is this a joke? This is empty except for one issue about an API deprecation. Very unhelpful information, and if that is a real roadmap, it is extremely disappointing that basically you consider copilot done. Nothing else to do there... If this is not a real roadmap, you are just making customers angrier... Again, I recommend looking into how you communicate with customers.

**Reply by mishmanners** [reply]
That is literally just the search for what's been announced and showcased on the public road for GitHub Copilot. The [Full Road Map](https://github.com/orgs/github/projects/4247) is very comprehensive. By no means does one issue mean GitHub Copilot is done, that's just what is announced _publicly_. You can also [check GitHub Next](https://githubnext.com/) for what the team might be working on Next. There's no public announcement on other platform support for GitHub Copilot chat. That's why I said keep an eye on the roadmap, the [GitHub blog](https://github.blog/), and [GitHub socials](https://twitter.com/github) for announcements.

**Reply by gshpychka** [reply]
>That is literally just the search for what's been announced and showcased on the public road for GitHub Copilot. The [Full Road Map](https://github.com/orgs/github/projects/4247) is very comprehensive

The full roadmap does not have any other issues on it related to Copilot, what do you mean by it being more comprehensive?

You mentioned that Copilot Chat in Neovim is coming soon - is that an official statement or just a guess?

**Reply by wootfish** [reply]
@mishmanners "There's no public announcement on other platform support for GitHub Copilot chat." I have to respectfully disagree, as in fact the following announcements and public statements have been made:

> [GitHub Copilot Chat now generally available for organizations and individuals](https://github.blog/2023-12-29-github-copilot-chat-now-generally-available-for-organizations-and-individuals/)

Note that this does not say "organizations and individuals who happen to be bought into the Visual Studio ecosystem".

> [All GitHub Copilot users can now enjoy natural language-powered coding with Copilot Chat at no additional cost.](https://github.blog/2023-12-29-github-copilot-chat-now-generally-available-for-organizations-and-individuals/)

Again, "All GitHub Copilot users" would seem to include Vim users. "No additional cost" excludes the cost of throwing away your organization's development environments and recentering around VS Code.

> [There’s also no need to install any additional extensions; Copilot chat extension is bundled together with the Copilot extension.](https://github.com/orgs/community/discussions/83931#discussioncomment-8131143)

This is apparently only true for _one_ of the many Copilot extensions.

> [You can start using GitHub Copilot Chat by installing the extension in your preferred environment.](https://docs.github.com/en/copilot/github-copilot-chat/using-github-copilot-chat-in-your-ide)

I would love to! Please either provide support or stop lying about it.

**Reply by wootfish** [reply]
There is a pattern here where you are acting as if this support is already in place, and this is problematic. I wish this support was in place, and some prospective customers will also wish it was, and rather than being misled I hope they can gain some balancing perspective by reading this thread (and the Copilot announcement thread).

**Reply by mishmanners** [reply]
I meant there's no more public announcements for more things coming up that aren't on the roadmap or already announced. Apologies for the confusion

**Reply by mishmanners** [reply]
I think we can take this as feedback and leave it here for the product team to see. I have provided you with all the information that is currently available. Sorry I couldn't do more.

**Reply by andresgutgon** [reply]
Thanks @mishmanners for your help and for your patience with us

**Reply by wootfish** [reply]
Yes, thank you for engaging and doing what you can. Anything you could do to raise this issue's visibility with the product team would be greatly appreciated.

**Reply by wvffle** [reply]
> I think we can take this as feedback and leave it here for the product team to see.

Yeah, and wait patiently until it's marked as stale and closed by a bot. Oh, wait... It's a discussion, it'll likely stay unresolved forever.

**Reply by gshpychka** [reply]
>I meant there's no more public announcements for more things coming up that aren't on the roadmap or already announced. 

Is your statement that Neovim support for Copilot Chat is coming soon a public announcement?

**Reply by tarfeef101** [reply]
is the mobile app seriously getting this before vim?? I know tons of people who use vim, i don't know anyone who does more than small PR reviews on their phone...

**Comment by z0rzi** [comment]
Hi there!

Over the past few days, I have been trying to put together a plugin, and I came up with something I'm decently happy about: https://github.com/z0rzi/ai-chat.nvim

I created it because I couldn't get `gptlang/CopilotChat.nvim` to work for me...

It's not tested, and I'm sure it could be improved in a lot of ways, but It does what I want it to do.
As it is right now, it probably won't work on windows...

If you feel like giving it a try, contributing, or raising issues, feel free to do it!

**Comment by neonvoidx** [comment]
https://github.com/CopilotC-Nvim/CopilotChat.nvim best we got at the moment

**Comment by GitMurf** [comment]
> https://github.com/CopilotC-Nvim/CopilotChat.nvim best we got at the moment

1. Can someone explain how this works?
2. Why does it require the python pre-reqs it mentions?
3. Is there some sort of internal api this is hacking into in order to access the chat functions?
4. Or is this doing some sort of hack like how you can write a code comment to get copilot to try and suggest something and this is doing some hackery in the background trying to get copilot to respond like a chat?

Thanks!

**Reply by kcuzner** [reply]
Looking at https://github.com/CopilotC-Nvim/CopilotChat.nvim/blob/main/rplugin%2Fpython3%2Fcopilot.py#L109 it seems like it's using a legit endpoint for copilot chat. In that file, it doesn't lie to the API that it's this bespoke plugin, so it doesn't seem illegitimate to me, other than if something breaks or changes in the API, Microsoft won't guarantee fixing it, as you'd expect with any 3rd party plugin.

**Reply by GitMurf** [reply]
> Looking at https://github.com/CopilotC-Nvim/CopilotChat.nvim/blob/main/rplugin%2Fpython3%2Fcopilot.py#L109 it seems like it's using a legit endpoint for copilot chat. In that file, it doesn't lie to the API that it's this bespoke plugin, so it doesn't seem illegitimate to me, other than if something breaks or changes in the API, Microsoft won't guarantee fixing it, as you'd expect with any 3rd party plugin.

@kcuzner thank you for your reply! so in other words it appears as if it is just using an undocumented API?

**Reply by GitMurf** [reply]
and is it fair to say the python pre-requisite is just because the developer prefers to write in python haha? and maybe lua doesn't have very good support for rest/curl/api type calls?

**Reply by MijikHna** [reply]
To be honest, I don't understand why Microsoft just don't provide proper API documentation for Copilot Chat. Vim community would then write and maintain a proper Copilot Chat plugin.

**Reply by gptlang** [reply]
Oh hey. Sorry to necro an old thread. Completely missed the activity as it occurred. I'm the original author of the prototype (It's now maintained by @deathbeam and @jellydn) 

> Can someone explain how this works?

I basically opened VSCode, fired up `mitmproxy`, and figured out the API endpoints from there.

> Why does it require the python pre-reqs it mentions?

Because I don't really know Lua. @deathbeam has since ported it to pure Lua without any Python dependencies

**Comment by jellydn** [comment]
Hi there,

I'm Dung Huynh, the maintainer of [CopilotC-Nvim/CopilotChat.nvim](https://github.com/CopilotC-Nvim/CopilotChat.nvim), and we're on a thrilling journey to integrate GitHub Copilot Chat with Neovim seamlessly. Our organization is dedicated to this vision, and we've made promising strides. But, we need *your* expertise and enthusiasm to propel us further! You could look at the open issues or discussions to know what we are working on.

**Why Join Us?**
- Contribute to a pioneering project blending Python and Lua for an enhanced Neovim experience.
- Be part of a vibrant community on our [Discord](https://discord.gg/vy6hJsTWaZ), where your ideas can spark innovation.
- Have a direct impact on features like [`clear chat buffer on new prompt`](https://github.com/CopilotC-Nvim/CopilotChat.nvim/pull/74) or [chat with your current buffer](https://github.com/CopilotC-Nvim/CopilotChat.nvim/pull/67). It's quite easy to add the feature you want with its flexible design. 

**Get Involved:**

Please look into our [documentation](https://github.com/CopilotC-Nvim/CopilotChat.nvim) for details on setup, usage, and how you can contribute, whether, through words, suggestions, etc.

**Contribute to a Future Where Coding Meets AI Seamlessly in Neovim!**

Warm regards,

Dung Huynh

**Reply by jellydn** [reply]
Hi everyone,

We are migrating to a Lua-based plugin on v2. Please take a try as this documentation https://github.com/CopilotC-Nvim/CopilotChat.nvim/blob/canary/MIGRATION.md and let us know if any issue.

**Reply by aeciolevy** [reply]
I could not make it work yet with Lazyvim.
 
![2024-03-07 16 25 59](https://github.com/community/community/assets/16165803/0616978b-2e8f-479a-b1a0-e185d8cfc604)

It was installed but I am not sure why.

The `CopilotChat` editor command is not found.

Nvim: 0.9.5
Plugin Manager: LazyVim

**Reply by aeciolevy** [reply]
My bad. Wrong branch!

**Reply by toddljones** [reply]
Great post. Thanks for sharing so thoughtfully.

**Comment by sala7khaled** [comment]


**Reply by peedrr** [reply]
@sala7khaled can i respectfully request that you edit your response?

Support for (neo)vim was promised. It now doesn't seem to be on any roadmap, let alone prioritised. People are understandably frustrated. So to describe the community's attempts at bridging the gap - a gap which MICROSOFT created - as "hacky", is really not OK imho.

A stable API would of course be great, however what we really want is an official implementation, not an API. M$ set the expectation and haven't met it. Your response is damaging for PR and reputation.

I think you've misread the room.

**Reply by GitMurf** [reply]
>... So to describe the community's attempts at bridging the gap - a gap which MICROSOFT created - as "hacky", is really not OK imho.

@peedrr I agree the message felt like there wasn't much value added and felt a little bit like "tough luck, you're on your own" BUT after I realized the very original message post in this thread (https://github.com/orgs/community/discussions/50939) actually used the term "hacky UI for Neovim", I then re-read the response and it did not seem as much in poor taste. That part should have been quoted to make it clear it was actually a response to the original post. Just wanted to provide that detail because I see several people reacting with emojis :)

**Reply by peedrr** [reply]
Fair point. I assumed it was referring to @jellydn 's efforts (i.e. the previous comment) 

The original issue text is nearly a year old!

**Reply by GitMurf** [reply]
> The original issue text is nearly a year old!

Yeah that is the part I still found a bit odd and also like what was the point of the message a year later and with actually no new information or clarity 🤷🏻‍♂️

I agree the message still seems a bit pointless (and frustrating) but it no longer seems as if in poor taste after reading the original post.

**Reply by JakobBruenker** [reply]
He also doesn't appear to be affiliated with github? It almost reads like an AI-generated response.

**Reply by neonvoidx** [reply]
> He also doesn't appear to be affiliated with github? It almost reads like an AI-generated response.

Generated via GitHub copilot chat plugin from neovims xd 😄

**Reply by queenofcorgis** [reply]
@sala7khaled 

We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by webel** [comment]
Am I missing some further context/update not shared here or why has the discussion died out? It's been 4 months since any "news".

**Comment by ArnoGW1** [comment]
Do you accept pull requests, if I get around to writing it? It can't be that hard for a user to add this to the official plugin. @Akash1134

**Comment by MissingRoberto** [comment]
Poking also about this. There should be feature parity between IDEs. Otherwise, do not say in the website that VIM is supported for Copilot.

**Comment by tarasowski** [comment]
Are there any valid alternatives to the chat feature (for vim)? e.g. from other providers?

**Comment by jorgejhms** [comment]
this one is the most recomended (Lazyvim is using it) [CopilotC-Nvim](https://github.com/CopilotC-Nvim/CopilotChat.nvim)

**Reply by tarasowski** [reply]
thx mate, but this doesn't work in vim, right?

**Reply by bulletmark** [reply]
@tarasowski unfortunately, for many reasons like this, it seems in-inevitable that all us vim users are going to have to port to nvim :(

**Reply by tarasowski** [reply]
@bulletmark I tested out some terminal apps, and I really like this one: https://github.com/kardolus/chatgpt-cli, you can send context with it by simply "cat file | chatgpt", also aider seems to be a valid option (similar to cursor.sh). Right now using chatgpt-cli.

**Reply by bulletmark** [reply]
@tarasowski my comment was not about this issue alone though. It seems to me that `nvim` has momentum which we can't ignore. More and more plugins (e.g. in Lua) will get created which we can't use in `vim` so we will have to jump on that bus ultimately. :(

**Reply by tarasowski** [reply]
@bulletmark okay, got your point. I like it to keep it simple and stupid, the only tool I'm missing is the full integration of copilot 🤖

**Reply by dimanyc** [reply]
> @tarasowski unfortunately, for many reasons like this, it seems in-inevitable that all us vim users are going to have to port to nvim :(

I've been trying to dodge the nvim hysteria for too long. It is, perhaps, time to embrace the inevitable...

**Comment by iufor** [comment]
Hello
The Copilot.vim plugin for Vim and Neovim offers GitHub Copilot's code suggestions but currently lacks native support for the Copilot Chat feature. While this chat function is integrated into tools like Visual Studio Code, it has not been officially extended to Copilot.vim. For interactive AI coding support in Vim environments, users might need alternative plugins or customized solutions​

**Reply by gptlang** [reply]
There is no official one but https://github.com/CopilotC-Nvim/CopilotChat.nvim/ works for Chat.

**Reply by atsiporu** [reply]
This one is Neovim only. The thesis stays: 
"Why there is no support for chat in the OFFICIAL plugin?"

**Reply by brianhuster** [reply]
> "Why there is no support for chat in the OFFICIAL plugin?"

Possibly you should ask the only 2 contributors of the plugin, `@tpope` and `@djensenius`

**Reply by gshpychka** [reply]
> > "Why there is no support for chat in the OFFICIAL plugin?"
> 
> Possibly you should ask the only 2 contributors of the plugin, `@tpope` and `@djensenius`

I believe it makes sense to ask the GitHub team in general. I doubt those contributors make the decisions regarding what to work on.

**Reply by llucax** [reply]
RIP copilot for *vim, the other editors keep getting more and more features, and this is stalled for ages now, it seems pretty clear that GitHub abandoned it.

**Reply by GitMurf** [reply]
> RIP copilot for *vim, the other editors keep getting more and more features, and this is stalled for ages now, it seems pretty clear that GitHub abandoned it.

I think people are missing the purpose / intent of copilot.vim. It is NOT copilot CHAT. It does what it is meant to which is provide autocompletions / code completions as you type your code.

That being said, yes I really hope GitHub invests in providing a Copilot Chat plugin and/or at least providing more api/sdk primitives that can be tapped into for others to implement some of the vscode Copilot Chat enhancements... but it is never going to be part of this copilot.vim plugin.

**Reply by bulletmark** [reply]
@GitMurf I don't sense anybody here who misunderstands the difference between copilot completions and chat. It's pretty clear what we are all asking for, given at the title of this discussion.

**Reply by atsiporu** [reply]
Because copilot.vim has been lacking those "chat" features for quite some time I found myself using vim-ai plugin more and more.
FWIW vim-ai plugin is farely simple and extendable and supports many LLMs including running local LLMs via Ollama.
Another option is to use natural vim visual mode substitution to the output of a running program which allows to do something like:
1. select some code
2. :'<,'>!ollama run <model> <ask it to do something with the code>
3. Get desired output.
The second step can be shortened if you create either scripts or source you bash functions that would be a short cut to call LLM with some predefined prompts. Which can turn your step two into: '<,'>!ollama-comment or '<,'>!ollama-review and etc.

**Comment by DanBradbury** [comment]
I was in need of chat for my own selfish purposes and wrote a plugin to mimic the functionality of [CopilotChat.nvim](https://github.com/CopilotC-Nvim/CopilotChat.nvim) in vimscript. Don't think this is the direction that github wants to go with exposing chat functionality but for the time being its working and I'm actively maintaining it to serve as a solution while we wait for official support.

https://github.com/DanBradbury/copilot-chat.vim

![](https://private-user-images.githubusercontent.com/2555073/420811068-0cd1119d-89c8-4633-972e-641718e6b24b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NjkwNjksIm5iZiI6MTc0NDQ2ODc2OSwicGF0aCI6Ii8yNTU1MDczLzQyMDgxMTA2OC0wY2QxMTE5ZC04OWM4LTQ2MzMtOTcyZS02NDE3MThlNmIyNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTJUMTQzOTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVhNTZmMTM4MjBjMDM1Mzg4MDBiYWI0YzIzMTlmYjk2YWM2ZjgzYjZjMmY3NzRjZTc3Y2IwZGVmNGM5ZDkwYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.x7Lh4kiPhUImZLPs8oFS5N2Di-YcSKlbLH7nbxWv50I)

**Reply by cesan3** [reply]
Thanks @DanBradbury, Fantastic solution!

**Reply by yermulnik** [reply]
@DanBradbury 👍🏻

**Reply by omercsp** [reply]
Excellent work @DanBradbury!

**Reply by atsiporu** [reply]
Yep this is fantastic plug-in. 
Thank you so much @DanBradbury .

You have inspired me and I migrated your vim script with the use of your plugin :) lol to bash. So full credit to you and to copilot itself! I was just a people person - I gave copilot plugin requests lol.

That gave me a command line version of copilot which I can now integrate into Linux tool pipelines.
I'm attaching a text file with bunch of bash functions. 
Which I source in my bashrc to get 
`copilot` command.
You can pipe data to copilot command while providing prompt as agreements.

This script requires location for device and chat token files, which I personally share with vim, or you can change the location and register another device by calling `copilot_chat_auth_get_bearer_token`

[bahrc-copilot.txt](https://github.com/user-attachments/files/20002599/bahrc-copilot.txt)
